common/archiver/s3store/historyArchiver.go

```diff
@@ -223,38 +267,61 @@ func (h *historyArchiver) Get(
 		}
 		token = &getHistoryToken{
 			CloseFailoverVersion: *highestVersion,
-			NextBatchIdx:         0,
 		}
 	}
 
-	key := constructHistoryKey(URI.Path(), request.DomainID, request.WorkflowID, request.RunID, token.CloseFailoverVersion)
-	encodedHistoryBatches, err := download(ctx, h.s3cli, URI, key)
+	prefix := constructHistoryKeyPrefixWithVersion(URI.Path(), request.DomainID, request.WorkflowID, request.RunID, token.CloseFailoverVersion)
+
+	results, err := h.s3cli.ListObjectsV2WithContext(ctx, &s3.ListObjectsV2Input{
+		Bucket:     aws.String(URI.Hostname()),
+		Prefix:     aws.String(prefix),
+		MaxKeys:    aws.Int64(int64(request.PageSize)), // Upper bound of history events
+		StartAfter: (*token).StartAfterKey,
+	})
+
 	if err != nil {
-		return nil, err
+		if isRetryableError(err) {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		return nil, &shared.BadRequestError{Message: err.Error()}
 	}
-	historyBatches, err := decodeHistoryBatches(encodedHistoryBatches)
-	if err != nil {
-		return nil, &shared.InternalServiceError{Message: err.Error()}
+	if len(results.Contents) == 0 {
+		return nil, &shared.BadRequestError{Message: errNoHistoryBatchesFound}
 	}
-	historyBatches = historyBatches[token.NextBatchIdx:]
 
 	response := &archiver.GetHistoryResponse{}
 	numOfEvents := 0
-	numOfBatches := 0
-	for _, batch := range historyBatches {
-		response.HistoryBatches = append(response.HistoryBatches, batch)
-		numOfBatches++
-		numOfEvents += len(batch.Events)
+	isTruncated := *results.IsTruncated
+	var lastKey *string
+	for _, item := range results.Contents {
 		if numOfEvents >= request.PageSize {
+			isTruncated = true
 			break
 		}
+		encodedRecord, err := download(ctx, h.s3cli, URI, *item.Key)
+		if err != nil {
+			if isRetryableError(err) {
+				return nil, &shared.InternalServiceError{Message: err.Error()}
+			}
+			return nil, &shared.BadRequestError{Message: err.Error()}
```

Comment:
At this stage, I don't think the error will be a badReqeustError.

Reply:
The reason I've been using BadRequestErrors here is that the client will retry otherwise. I'm of course open to suggestions on how you would like it to be handled

Reply:
Yeah, we didn't handle retry very well on our client side. It will retry on internal service error, but not bad request error or entity not exist error. 

But when returning an error here, I think the most important thing is the error should make sense to our user. Returning a badRequestError will make user confuse as their request is totally valid. So I think whether or not the request should be retried should be handled by some other logic at the client side, not here. 

Ideally we should have a nonRetryableInternalServiceError type, but for now, I guess retrying a few time is ok.

Reply:
Thanks, however now after the refactor EnityNotExists could appear here as is confirmed by the TestGet_Fail_KeyNotExist unit test. Otherwise it will return InternalServiceError

Reply:
Yeah, if it's entity not exist, then return `shared.EntityNotExistError{}`.

Reply:
Good. I think all comments should be adressed now